stan/math/prim/functor/hcubature.hpp

```diff
@@ -62,314 +67,279 @@ static constexpr double gwd7[4] = {1.2948496616886969327061143267787e-01,
-    std::vector<double> temp(dim, 0.0);
+  Eigen::Matrix<Scalar, Eigen::Dynamic, Eigen::Dynamic> p
+      = Eigen::MatrixXd::Zero(dim, choose_dimk * std::pow(2, k));
+  int current_col = 0;
+  const auto inner_iter_len = std::pow(2, k);
+  std::vector<bool> index;
+  index.reserve(inner_iter_len * (choose_dimk + 1));
+  for (int i = 1; i != choose_dimk + 1; i++) {
     combination(c, dim, k, i);
-    std::vector<bool> index;
-    index.clear();
-    for (std::size_t j = 0; j != std::pow(2, k); j++) {
-      increment(index, k, lambda, c, temp);
-      p.push_back(temp);
+    index.push_back(false);
+    p(c.array() - 1.0, current_col).setConstant(lambda);
+    current_col += 1;
+    for (int j = 1; j != inner_iter_len; j++, current_col++) {
+      p.col(current_col) = p.col(current_col - 1);
+      int first_zero
+          = std::distance(std::begin(index),
+                          std::find(std::begin(index), std::end(index), false));
+      const std::size_t index_size = index.size();
+      if (first_zero == index_size) {
+        index.flip();
+        p(c.segment(0, index.size()).array() - 1, current_col).array() *= -1;
+        index.push_back(true);
+        p(c[index.size() - 1] - 1, current_col) = -lambda;
+      } else {
+        for (int h = 0; h != first_zero + 1; h++) {
+          index[h].flip();
+        }
+        p(c.segment(0, first_zero + 1).array() - 1, current_col).array() *= -1;
+      }
     }
+    index.clear();
   }
+  return p;
 }
 
 /**
  * Compute the integral of the function to be integrated (integrand) from a to b
  * for one dimension.
  *
  * @tparam F type of the integrand
- * @tparam T_pars type of the parameters for the integrand
+ * @tparam T_a type of lower limit of integration
+ * @tparam T_b type of upper limit of integration
+ * @tparam ParsPairT type of the pair of parameters for the integrand
  * @param integrand function to be integrated
  * @param a lower limit of integration
  * @param b upper limit of integration
- * @param pars parameters for the integrand
+ * @param pars_pair Pair of parameters for the integrand
  * @return numeric integral of the integrand and error
  */
-template <typename F, typename T_pars>
-std::tuple<double, double> gauss_kronrod(const F& integrand, const double& a,
-                                         const double& b, T_pars& pars) {
-  std::vector<double> c(1, 0);
-  std::vector<double> cp(1, 0);
-  std::vector<double> cm(1, 0);
-  c[0] = 0.5 * (a + b);
-  double delta = 0.5 * (b - a);
-  double f0 = integrand(c, pars);
-  double I = f0 * wd7[7];
-  double Idash = f0 * gwd7[3];
-  for (std::size_t i = 0; i != 7; i++) {
-    double deltax = delta * xd7[i];
-    cp[0] = c[0] + deltax;
-    cm[0] = c[0] - deltax;
-    double fx = integrand(cp, pars);
-    double temp = integrand(cm, pars);
-    fx += temp;
+template <typename F, typename T_a, typename T_b, typename ParsPairT>
+inline auto gauss_kronrod(const F& integrand, const T_a a, const T_b b,
```

Comment:
Is there a reason not to just use the `gauss_kronod` that exists in boost already? I think it would be better to use their version instead of rolling our own

https://live.boost.org/doc/libs/1_84_0/libs/math/doc/html/math_toolkit/gauss_kronrod.html

Reply:
The reason is that we took the implementation by Steven G. Johnson and translated it from julia to c++ to Stan.

Reply:
I thought every function that we use should be in Stan. If this is not the case and if the boost implementation does the same things and results are the same, we can use boost. I didn't know the boost implementation before.
But this sentence in the boost documentation gives reason to not use this implementation: 'These routines are deliberately kept relatively simple: when they work, they work very well and very rapidly. However, no effort has been made to make these routines work well with end-point singularities or other "difficult" integrals.'
I would say our integrals are quite 'difficult' and complex.

Also does this implementation need different input arguments than our implementation.

If you think this does not matter, you can change the code to use the boost implementation.

Reply:
> I thought every function that we use should be in Stan.

If we have something available that's been battle tested then it's a better choice than having our own implementation. 

> But this sentence in the boost documentation gives reason to not use this implementation: 'These routines are deliberately kept relatively simple: when they work, they work very well and very rapidly. However, no effort has been made to make these routines work well with end-point singularities or other "difficult" integrals.'
I would say our integrals are quite 'difficult' and complex.

Yes but that assumes our own implementation of gauss kronrod is better at the integrals we want. We don't have any evidence of that.

> If you think this does not matter, you can change the code to use the boost implementation.

I'll see if I have time this week. I think it's just changing where we call gauss kronrod like 

```
Scalar result;
Scalar err;
auto gk_lambda = [&integrand, &pars](auto&& c) {
  return stan::math::apply([](auto&& integrand, auto&& c, auto&&... args) {
    return integrand(c, args...);
  }, pars, integrand, c);
};
result = boost::gauss_kronrod::integrate(gk_lambda, a[0], b[0], 15, boost::tools::root_epsilon<Real>(), &err);
```

Reply:
Hi @SteveBronder, your last comment helped me to change the function call myself. I inserted this gk_lambda and deleted our gauss_kronrod, our gauss_kronrod calls and the three helper arrays at the beginning of the document. Can you have a look at this?

One of the three new gauss_kronrod calls look like this:
`    result = gauss_kronrod<Scalar, 15>::integrate(gk_lambda, a[0], b[0], 15, 0,
                                                  &err);`
Are the numbers correct: `<Scalar, 15>`, `..., 15, 0, &err)` ?

Value-tests run successfully.

Reply:
hm ... but the mix tests do not seem to work with this implementation in this form ... Do you have an idea?

Reply:
Can you write a test comparing the output of doubles between the two? My guess is that you need to change the depth for the boost version or one of the other parameters for a more precise estimate

Reply:
The test  `test/unit/math/prim/prob/hcubature.hpp` works without errors. So the `double` inputs are fine.
The error message says  `error: the type ‘const std::array<stan::math::var_value<double>, 8>’ of ‘constexpr’ variable ‘data’ is not literal`. It seems that gauss_kronrod cannot work with the other input types, doesn't it?

Reply:
That's from the points though right? We shouldn't need those points if we are using boosts gauss kronrod since those values are made in boosts impl. I can look at this tmrw